ls1intum/Artemis
# `Lectures`: Allow instructors to define relations between learning goals

### Checklist
#### General


#### Server
#### Client
- [ ] I translated all newly inserted strings into English and German. _(Did not translate admin-only interface.)_

### Motivation and Context



Working towards integrating Adaptive Learning in Artemis, instructors need to be able to define relations between learning goals. This will later allow Artemis to generate Learning Paths between the individual learning goals.

### Description


We model a new relation between two learning goals that describes their relation on a graph-like structure (you can imagine directed edges in a graph.). Additionally, a new attribute is added to learning goals to persist their [taxonomy](https://en.wikipedia.org/wiki/Bloom%27s_taxonomy).
To make this as versatile as possible, the relation is defined through a new class called `LearningGoalRelation` that holds the tail learning goal, head learning goal and the relation type:

- RELATES: A generic relation between two learning goals.
- ASSUMES: The tail learning goal assumes that the student already achieved the head learning goal.
- EXTENDS: The tail learning goal extends the head learning goal on the same topic in more detail.
- MATCHES: The tail learning goal matches the head learning goal (e.g., a duplicate).

To allow you to test this feature, we implemented an admin-only interface with which you to create relations, display them in a graph and remove them again (by clicking on them in the graph).

<img src="https://user-images.githubusercontent.com/102237363/185746409-0a7d7ea6-a912-4049-a6c4-c3d8641af540.png" width="600" />

### Steps for Testing

Prerequisites:
- 1 Admin

1. Log in to Artemis as admin
2. Navigate to Course Administration
3. Go to the Learning Goal Management page
4. Create some learning goals
5. Add some relations between the learning goals and make sure they are persisted
7. Remove a relation by clicking on it in the graph

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Method | Line |
|------------|-------:|-----:|
| LearningGoalService.java | 100% | 96% |
| LearningGoalResource.java | 100% | 91% |

### Screenshots



<img width="800" alt="Screenshot 2022-08-30 at 17 21 25" src="https://user-images.githubusercontent.com/102237363/187476385-924f1d43-9468-4c82-b59d-20dfe93a52c0.png">
